test/runtests.jl

```diff
@@ -12,28 +19,71 @@ struct Foo
 end
 
 # test basic match
-x = nothing
-@test (@match Foo(x,2) = Foo(1,2)) == Foo(1,2)
-@test x == 1
+begin
+    local x = nothing
+    @test (@match Foo(x,2) = Foo(1,2)) == Foo(1,2)
+    @test x == 1
+end
 
 # variables not bound if match fails
-x = nothing
-@test_throws MatchFailure @match Foo(x, 3) = Foo(1,2)
-@test x == nothing
+begin
+    local x = nothing
+    @test_throws MatchFailure @match Foo(x, 3) = Foo(1,2)
+    @test x == nothing
+end
 
 # doesn't overwrite variables in outer scope
-x = nothing
-@test (@match Foo(1,2) begin
-    Foo(x,2) => x
-end) == 1
-@test x == nothing
+begin
+    local x = nothing
+    @test (@match Foo(1,2) begin
+        Foo(x,2) => x
+    end) == 1
+    @test x == nothing
+end
 
 # variables not bound if guard fails
-x = nothing
-@test_throws MatchFailure @match Foo(1,2) begin
-  Foo(x, 2) where x != 1 => :ok
+begin
+    local x = nothing
+    @test_throws MatchFailure @match Foo(1,2) begin
+      Foo(x, 2) where x != 1 => :ok
+    end
+    @test x == nothing
+end
+
+# match one struct field by name
+begin
+    local x1 = nothing
+    @test (@match Foo(1,2) begin
+           Foo(x=x1) => x1
+    end) == 1
+end
+
+# match struct with mix of by-value and by-field name
+begin
+    local x1 = nothing
+    @test (@match Foo(1,2) begin
+           Foo(0,2) => nothing
+           Foo(x=x1) => x1
+    end) == 1
+end
+
+# match multiple struct fields by name
+begin
+    local x1 = nothing
+    local y1 = nothing
+    @test (@match Foo(1,2) begin
+           Foo(x=x1,y=y1) => (x1,y1)
+    end) == (1,2)
+end
+
+# match struct field by name redundantly
+begin
+    local x1 = nothing
+    local x2 = nothing
+    @test_throws assertion_error @eval @match Foo(1,2) begin
+           Foo(x=x1,x=x2) => (x1,x2)
+    end
```

Comment:
Can you also add one test in here that checks the struct fields against known values, rather than capturing them?

eg:
```julia
@test (@match Foo(1,2) begin
           Foo(x=x1,y=2) => (x1, 2)
    end) == 1,2
@test x1 == 1
```